### PR TITLE
Silently ignored networking.nameservers setting #29202

### PR DESCRIPTION
**Pull request by gwitmond**
It's better to break with a comment than to silently ignore a user's setting.

###### Motivation for this change

The networking configuration set the resolver configuration (/etc/resolv.conf) to use 127.0.0.1 as the nameserver when the services.bind.enable is set to true. Or when services.dnsmasq.resolveLocalQueries is enabled.

This PR adds an assertion to verify if the user set the networking.nameservers options when either service is enabled. This makes it clear that the user cannot set the nameservers for the resolver.

###### Things done

- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux

- [ ] Tested according to 'steps to reproduce' in issue #29202.`

### CODE REVIEW

File: nixos/modules/config/networking.nix [hunk 1]
```diff
@@ -9,7 +9,7 @@ let
   cfg = config.networking;
   dnsmasqResolve = config.services.dnsmasq.enable &&
                    config.services.dnsmasq.resolveLocalQueries;
-  hasLocalResolver = config.services.bind.enable || dnsmasqResolve;
+  hasLocalResolver = cfg.nameservers == [] && ( config.services.bind.enable || dnsmasqResolve );
```
Comment (peti) [inline nixos/modules/config/networking.nix]: What exactly happens here when `nameservers` is *not* empty?

Comment (vcunat) [reply nixos/modules/config/networking.nix]: The intention of this PR (apparently): only use nameservers from the explicit `nameservers` list and not `127.0.0.1`.

Side note: I added another resolver months ago and didn't realize we had this magic; I should amend `hasLocalResolver` for that.

Comment (gwitmond) [reply nixos/modules/config/networking.nix]: @peti: When 'nameservers' is empty, it follows the old behaviour. That is: it sets 'nameserver = 127.0.0.1' in networking.nix when either services.bind.enable == true or services.dnsmasq.resolveLocalQueries == true at https://github.com/gwitmond/nixpkgs/blob/4b8c7065e2d9f2961855e7ed2cd51b4bbb6d5f11/nixos/modules/config/networking.nix#L241

If neither networking.nameservers is set nor either of those services, it defaults to standard behaviour.

@vcunat: Indeed, my goal is to give the user specified nameservers priority over any automatic ones.

**Review by peti — CHANGES_REQUESTED** [review]
I actually have a machine that has `nameservers` configured which aren't used because there is a local BIND running, and I do rely on the fact that the local bind overrides the other setting. If we'd merge this PR, then this would break my server's setup. Therefore, I'm 👎 on this PR.

I think it would be okay to add an `assertion` that catches such an arguably inconsistent configuration, but silently changing the behavior of NixOS sounds like a bad idea to me.

**Comment by gwitmond** [comment]
Ignore this PR for now, as I added a bunch of garbage...

**Comment by gwitmond** [comment]
I've created a new PR at https://github.com/NixOS/nixpkgs/pull/29503

**Pull request closed without merging**